File: Tim/Commands/Utility/InteractionControls.java
package Tim.Commands.Utility;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;

import Tim.Commands.ICommandHandler;
import Tim.Data.ChannelInfo;
import Tim.Data.CommandData;
import Tim.Tim;
import org.pircbotx.hooks.types.GenericMessageEvent;
import org.pircbotx.hooks.types.GenericUserEvent;

public class InteractionControls implements ICommandHandler {
	private HashSet<String> interactions = new HashSet<>();

	public InteractionControls() {
		interactions.addAll(ChannelInfo.getChatterDefaults()
									   .keySet());
		interactions.addAll(ChannelInfo.getCommandDefaults()
									   .keySet());
		interactions.add("hugs");
		interactions.remove("bored");
		interactions.remove("chainstory");
		interactions.remove("commandment");
		interactions.remove("dance");
		interactions.remove("dice");
		interactions.remove("eightball");
		interactions.remove("expound");
		interactions.remove("markov");
		interactions.remove("ping");
		interactions.remove("sing");
		interactions.remove("woot");
	}

	public static void helpSection(GenericUserEvent event) {
		String[] strs = {
			"Interaction Control Command:",
			"    This command allows you to control the ways Timmy interacts with you, in case you don't enjoy certain of his behaviors.",
			"    !interactionflag - Command for viewing and controlling the interactions.",
			};

		for (String str : strs) {
			event.getUser()
				 .send()
				 .message(str);
		}
	}

	public static boolean interactWithUser(String username, String interaction) {
		if (!Tim.db.userInteractionSettings.containsKey(username.toLowerCase())) {
			return true;
		}

		HashMap<String, Boolean> userSettings = Tim.db.userInteractionSettings.get(username.toLowerCase());
		if (!userSettings.containsKey(interaction)) {
			return true;
		}

		return userSettings.get(interaction);
	}

	@Override
	public boolean handleCommand(CommandData commandData) {
		if (commandData.type != CommandData.CommandType.TIMMY_USER) {
			return false;
		}

		switch (commandData.command) {
			case "interactionflag":
				if (checkVerified(commandData.getMessageEvent())) {
					if (commandData.args.length == 1 && commandData.args[0].equalsIgnoreCase("list")) {
						String target = commandData.issuer;
						if (Tim.db.userInteractionSettings.containsKey(target.toLowerCase())) {
							HashMap<String, Boolean> data = Tim.db.userInteractionSettings.get(target.toLowerCase());

							commandData.event.respond("Sending status of interaction settings via private message.");

							data.keySet()
								.forEach((setting) -> commandData.getUserEvent()
																 .getUser()
																 .send()
																 .message(setting + ": " + data.get(setting)
																							   .toString()));
						} else {
							commandData.event.respond("No settings stored for that user.");
						}
					} else if (commandData.args.length == 3 && commandData.args[0].equalsIgnoreCase("set")) {
						String target = commandData.issuer;
						if (!Tim.db.userInteractionSettings.containsKey(target.toLowerCase())) {
							Tim.db.userInteractionSettings.put(target.toLowerCase(), new HashMap<>());
						}

						HashMap<String, Boolean> userData = Tim.db.userInteractionSettings.get(target.toLowerCase());

						boolean flag = false;
						if (!"0".equals(commandData.args[2])) {
							flag = true;
						}

						if (commandData.args[1].equalsIgnoreCase("all")) {
							for (String key : sortedInteractions()) {
								userData.put(key, flag);
							}

							Tim.db.saveInteractionSettings();

							commandData.event.respond("All interaction flags updated.");
						} else {
							if (interactions.contains(commandData.args[1])) {
								if (!flag) {
									userData.put(commandData.args[1], false);
								} else {
									userData.remove(commandData.args[1]);
								}

								Tim.db.saveInteractionSettings();

								commandData.event.respond("Interaction flag updated.");
							} else {
								commandData.event.respond("I'm sorry, but I don't have a setting for " + commandData.args[1]);
							}
						}
					} else {
						commandData.event.respond("Usage: !interactionflag list OR !interactionflag set <type> <0/1>");
						commandData.event.respond("Valid Interaction Types: all, " + String.join(", ", sortedInteractions()));
					}
				} else {
					commandData.event.respond("You must be logged in via NickServ to use these commands.");
				}

				return true;
		}

		return false;
	}

	private boolean checkVerified(GenericMessageEvent event) {
		return event.getUser() != null && event.getUser()
											   .isVerified();
	}

	private ArrayList<String> sortedInteractions() {
		ArrayList<String> sorted = new ArrayList<>(interactions);
		Collections.sort(sorted);
		return sorted;
	}
}

File: timmy/command_processors/utility/interactioncontrols.py
from irc.dict import IRCDict

from timmy import core, db_access
from timmy.command_processors.base_command import BaseCommand
from timmy.data.channel_data import ChannelData
from timmy.data.command_data import CommandData
from timmy.data.userdata import UserData
from timmy.db_access import user_directory


class InteractionControls(BaseCommand):
    user_commands = ['interactionflag']

    help_topics = [('user', 'interaction controls', '!interactionflag', 'Command for managing how Timmy will interact '
                                                                        'with you, randomly or via amusement commands.')
                   ]

    def __init__(self):
        self.interaction_settings = IRCDict()
        self.initialized = False
        self.interactions = set()

    def late_init(self):
        super().__init__()

    def init(self):
        if not self.initialized:
            for entry in ChannelData.chatter_settings_defaults['types'].keys():
                self.interactions.add(entry)

            for entry in ChannelData.command_defaults.keys():
                self.interactions.add(entry)

            self.interactions.add('hugs')
            self.interactions.remove('bored')
            self.interactions.remove('chainstory')
            self.interactions.remove('commandment')
            self.interactions.remove('dance')
            self.interactions.remove('dice')
            self.interactions.remove('eightball')
            self.interactions.remove('expound')
            self.interactions.remove('markov')
            self.interactions.remove('ping')
            self.interactions.remove('sing')
            self.interactions.remove('woot')

            self._load_interaction_settings()

            self.initialized = True

    def process(self, command_data: CommandData) -> None:
        self.init()

        if core.user_perms.is_registered(command_data.issuer):
            target: UserData = user_directory.find_user_data(command_data.issuer)

            if command_data.arg_count == 1 and command_data.args[0] == 'list':
                if target.authed_user in self.interaction_settings:
                    data: dict = self.interaction_settings[target.authed_user]

                    if not command_data.in_pm:
                        self.respond_to_user(command_data, "Sending status of interaction settings via private "
                                                           "message.")

                    from timmy.core import bot_instance
                    for key, value in data.items():
                        bot_instance.connection.privmsg(command_data.issuer, "{}: {}".format(key, value))
                else:
                    self.respond_to_user(command_data, "No settings stored for that user.")

            elif command_data.arg_count == 3 and command_data.args[0].lower() == 'set':
                if target.authed_user not in self.interaction_settings:
                    self.interaction_settings[target.authed_user] = IRCDict()

                data = self.interaction_settings[target.authed_user]
                flag = '1' == command_data.args[2]

                if command_data.args[1].lower() == 'all':
                    for key in self.interactions:
                        data[key] = flag

                    self._save_interaction_settings()

                    self.respond_to_user(command_data, "All interaction flags updated.")
                else:
                    if command_data.args[1] in self.interactions:
                        # We only need to store disabled interactions. Default is enabled.
                        if not flag:
                            data[command_data.args[1]] = flag
                        else:
                            del (data[command_data.args[1]])

                        self._save_interaction_settings()

                        self.respond_to_user(command_data, "Interaction flag updated.")
                    else:
                        self.respond_to_user(command_data, f"I'm sorry, but I don't have a setting for "
                                                           f"{command_data.args[1]}")
            else:
                self.respond_to_user(command_data, "Usage: !interactionflag list OR !interactionflag set <type> <0/1>")
                self.respond_to_user(command_data, "Valid interaction types: all, {}"
                                     .format(", ".join(self.interactions)))
        else:
            self.respond_to_user(command_data, "You must be logged in with NickServ to use these commands.")

    def interact_with_user(self, username: str, interaction: str) -> bool:
        self.init()

        target: UserData = user_directory.find_user_data(username)

        if target is None or target.authed_user not in self.interaction_settings:
            return True

        if interaction not in self.interaction_settings[target.authed_user]:
            return True

        if core.user_perms.is_ignored(target, 'any'):
            return False

        return self.interaction_settings[target.authed_user][interaction]

    def _save_interaction_settings(self) -> None:
        truncate_statement = "TRUNCATE `user_interaction_settings`"
        insert_statement = "INSERT INTO `user_interaction_settings` SET `username` = %(username)s, " \
                           "`setting` = %(setting)s, `value` = %(value)s"

        conn = db_access.connection_pool.get_connection()
        cursor = conn.cursor()
        cursor.execute(truncate_statement)

        for user, settings in self.interaction_settings.items():
            for setting, value in settings.items():
                cursor.execute(
                        insert_statement, {
                            'username': "{}".format(user),
                            'setting':  "{}".format(setting),
                            'value':    value == 1
                        }
                )

        db_access.connection_pool.close_connection(conn)

    def _load_interaction_settings(self) -> None:
        select_statement = "SELECT `username`, `setting`, `value` FROM `user_interaction_settings`"

        conn = db_access.connection_pool.get_connection()
        cursor = conn.cursor(dictionary = True)
        cursor.execute(select_statement)

        for row in cursor:
            if row['username'] not in self.interaction_settings:
                self.interaction_settings[row['username']] = IRCDict()

            self.interaction_settings[row['username']][row['setting']] = row['value'] == 1

        db_access.connection_pool.close_connection(conn)
